From 149e1dbf856046adb6448fc88dc31c2e20251495 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Wed, 23 Sep 2026 23:25:43 -0400
Subject: [PATCH 02/13] ip/opus: fix opusfile callbacks and fdsan abort on
 invalid files

Previously, op_fdopen would overwrite our callbacks with the opusfile
stdio ones (cb is an output parameter), and fdopen would take ownership
of the fd, making fdsan abort due to it being closed with close (rather
than fdclose) (and also leaking the FILE) when an invalid file is open.

Fixing the callbacks also means we need to fix the seek callback since
opusfile requires the return to be success/failure, rather than the new
offset.

The only notable behaviour difference between our callbacks and the
opusfile stdio ones are the read buffering behaviour. Since opusfile
reads 2KiB chunks, and most libc implementations (glibc, bionic) use a
st_blksize-size buffer (usually 4KiB), this doubles the number of read
syscalls. For musl, the syscall count is unchanged since it reads
directly, and only uses an internal 1KiB buffer for overflow. At any
rate, most other plugins (e.g., vorbis, flac) also use raw i/o too, so
it's not really a big deal.

diff --git a/ip/opus.c b/ip/opus.c
index 775b647..36ae70c 100644
--- a/ip/opus.c
+++ b/ip/opus.c
@@ -46,7 +46,7 @@ static int read_func(void *datasource, unsigned char *ptr, int size)
 static int seek_func(void *datasource, opus_int64 offset, int whence)
 {
 	struct input_plugin_data *ip_data = datasource;
-	return lseek(ip_data->fd, offset, whence);
+	return lseek(ip_data->fd, offset, whence) == -1 ? -1 : 0;
 }
 
 static int close_func(void *datasource)
@@ -66,7 +66,7 @@ static opus_int64 tell_func(void *datasource)
 	return lseek(ip_data->fd, 0, SEEK_CUR);
 }
 
-static OpusFileCallbacks callbacks = {
+static const OpusFileCallbacks callbacks = {
 	.read = read_func,
 	.seek = seek_func,
 	.tell = tell_func,
@@ -77,19 +77,12 @@ static int opus_open(struct input_plugin_data *ip_data)
 {
 	struct opus_private *priv;
 	int rc;
-	void *source;
 
 	priv = xnew(struct opus_private, 1);
 	priv->current_link = -1;
 	priv->of = NULL;
 
-	source = op_fdopen(&callbacks, ip_data->fd, "r");
-	if (source == NULL) {
-		free(priv);
-		return -IP_ERROR_INTERNAL;
-	}
-
-	priv->of = op_open_callbacks(source, &callbacks, NULL, 0, &rc);
+	priv->of = op_open_callbacks(ip_data, &callbacks, NULL, 0, &rc);
 	if (rc != 0) {
 		d_print("op_open_callbacks failed: %d:%s\n", rc, strerror(rc));
 		free(priv);
